gen.go

```diff
@@ -1019,6 +1117,178 @@ func emitCborUnmarshalSliceField(w io.Writer, f Field) error {
+				return fmt.Errorf("{{ .Name }}: byte array too large (%d)", extra)
+			}
+			if maj != cbg.MajByteString {
+				return fmt.Errorf("expected byte array")
+			}
+	{{ if .PreserveNil }}
+			{{ .Name }} = make({{ .TypeName }}, extra)
+	{{ else }}
+			if extra > 0 {
+				{{ .Name }} = make({{ .TypeName }}, extra)
+			}
+	{{ end }}
+			if _, err := io.ReadFull(cr, {{ .Name }}[:]); err != nil {
+				return err
+			}
+
+	{{ if .PreserveNil }}
+		}
+	}
+	{{ end }}
+`)
+	}
+
+	if err := doTemplate(w, f, `
+	if extra > {{ MaxLen .MaxLen "cbg.MaxLength" }} {
+		return fmt.Errorf("{{ .Name }}: array too large (%d)", extra)
+	}
+`); err != nil {
+		return err
+	}
+
+	err = doTemplate(w, f, `
+	if maj != cbg.MajArray {
+		return fmt.Errorf("expected cbor array")
+	}
+	{{ if .PreserveNil }}
+	{{ .Name }} = make({{ .TypeName }}, extra)
+	{{ else }}
+	if extra > 0 {
+		{{ .Name }} = make({{ .TypeName }}, extra)
+	}
+	{{ end }}
+	for {{ .IterLabel }} := 0; {{ .IterLabel }} < int(extra); {{ .IterLabel }}++ {
+`)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(w, "\t\t{\n\t\t\tvar maj byte\n\t\tvar extra uint64\n\t\tvar err error\n")
+	fmt.Fprintf(w, "\t\t\t_ = maj\n\t\t_ = extra\n\t\t_ = err\n")
+
+	switch e.Kind() {
+	case reflect.Struct:
+		subf := Field{
+			Type:    e,
+			Pkg:     f.Pkg,
+			Pointer: pointer,
+			Name:    f.Name + "[" + f.IterLabel + "]",
+		}
+		err := emitCborUnmarshalStructField(w, subf)
+		if err != nil {
+			return err
+		}
+	case reflect.Uint64:
+		subf := Field{
+			Type: e,
+			Pkg:  f.Pkg,
+			Name: f.Name + "[" + f.IterLabel + "]",
+		}
+		err := emitCborUnmarshalUint64Field(w, subf)
+		if err != nil {
+			return err
+		}
+	case reflect.Int64:
+		subf := Field{
+			Type: e,
+			Pkg:  f.Pkg,
+			Name: f.Name + "[" + f.IterLabel + "]",
+		}
+		err := emitCborUnmarshalInt64Field(w, subf)
+		if err != nil {
+			return err
+		}
+	case reflect.Array:
+		nextIter := string([]byte{f.IterLabel[0] + 1})
+		subf := Field{
+			Name:      fmt.Sprintf("%s[%s]", f.Name, f.IterLabel),
+			Type:      e,
+			IterLabel: nextIter,
+			Pkg:       f.Pkg,
+		}
+		if err := emitCborUnmarshalArrayField(w, subf); err != nil {
+			return err
+		}
+	case reflect.Slice:
+		nextIter := string([]byte{f.IterLabel[0] + 1})
+		subf := Field{
+			Name:      fmt.Sprintf("%s[%s]", f.Name, f.IterLabel),
+			Type:      e,
+			IterLabel: nextIter,
+			Pkg:       f.Pkg,
+		}
+		if err := emitCborUnmarshalSliceField(w, subf); err != nil {
+			return err
+		}
+
+	case reflect.String:
+		subf := Field{
+			Type: e,
+			Pkg:  f.Pkg,
+			Name: f.Name + "[" + f.IterLabel + "]",
+		}
+		err := emitCborUnmarshalStringField(w, subf)
+		if err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("do not yet support slices of %s yet", e.Elem())
+	}
+
+	if err := doTemplate(w, f, `
+	{{ if .PreserveNil }}
+				}
+			}
+	{{ end }}
+		}
+	}
+	`); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func emitCborUnmarshalArrayField(w io.Writer, f Field) error {
```

Comment:
Ditto. Can we avoid duplicating this logic by assuming that preservenil implies slice and letting the user deal with the compile error?